### PR TITLE
Optimize permute gemm onednn

### PR DESCRIPTION
**Pull request by hyunback**
We can optimize out permute in permute+gemm  or gemm+permute pattern in dGPU.
Permute in and out should be compatible  and no post-op. After opt out permute, we need to pass updated tag corresponding to permute_order to oneDNN.

### Tickets:
 - *104826*

### CODE REVIEW

File: src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp [hunk 1]
```diff
@@ -526,3 +527,80 @@ INSTANTIATE_TEST_SUITE_P(
...
+    }
+};
+
+class gemm_2in_permute : public GemmFusingTestOneDNN {};
+TEST_P(gemm_2in_permute, basic) {
+    auto p = GetParam();
+    create_topologies(
+        input_layout("input0", get_input_layout(p, 0)),
+        input_layout("input1", get_input_layout(p, 1)),
+        gemm("gemm_prim", { input_info("input0"), input_info("input1") }, data_types::f16),
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp:590]: could you add test case for permute-gemm case as well?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 2]
```diff
@@ -1873,9 +1873,58 @@ void layout_optimizer::select_preferred_formats_for_onednn(program_node& node, d
...
+        if (node.is_type<gemm>()) {
+            for (size_t idx = 0 ; idx < node.get_dependencies().size() ; idx++) {
+                if (node.get_dependency(idx).is_type<permute>()) {
+                    auto& pnode = node.get_dependency(idx);
+                    if (pnode.has_fused_primitives()) {
+                        continue;
+                    }
+                    auto input_lay = pnode.get_dependency(0).get_output_layout();
+                    auto output_lay = pnode.get_output_layout();
+                    if (input_lay.compatible(output_lay)) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1886]: from my experience, there was a corner case where perf drop was observed. I think it is risky to merge any permute. How do you think?

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp (hunk 1 above)
Comment (hyunback) [reply src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp:590]: Applied.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 2 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1893]: Let's check the performance model by mode, and discuss it

File: src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp [hunk 3]
```diff
@@ -82,6 +82,8 @@ struct format {
         byxf,                                   ///< used in bitmaps, input from user i.e b images of RGB format
         fyxb,                                   ///< format not used inside clDNN, but supported in reorder as extension
         bzyxf,
+        byfx,
+        bxfy,
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp]: Applied

File: src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp [hunk 4]
```diff
@@ -526,3 +527,80 @@ INSTANTIATE_TEST_SUITE_P(
...
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    fusings_gpu, gemm_2in_permute, ::testing::ValuesIn(std::vector<gemm_test_params>{
+        gemm_test_params{CASE_GEMM_2IN_FP16_1, 3, 4},
+        gemm_test_params{CASE_GEMM_2IN_FP16_2, 3, 4},
+        gemm_test_params{CASE_GEMM_2IN_FP16_3, 3, 4},
+    }));
+
+#endif // ENABLE_ONEDNN_FOR_GPU
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/tests/unit/fusions/gemm_fusion_test.cpp]: Applied

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 5]
```diff
@@ -1873,9 +1873,58 @@ void layout_optimizer::select_preferred_formats_for_onednn(program_node& node, d
             GPU_DEBUG_LOG << "select_preferred_formats:" << node.id() << ": " << fmt_to_str(target_format) << " --> " << fmt_to_str(target_format)
                           << " For index : " << idx << std::endl;
         }
+        // Optimized out permute from permute-gemm pattern. e.f permute -> gemm
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Applied

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 6]
```diff
@@ -1873,9 +1873,58 @@ void layout_optimizer::select_preferred_formats_for_onednn(program_node& node, d
...
+                }
+            }
+            // gemm -> permute
+            if (node.get_users().size() == 1 && node.get_users().front()->is_type<permute>() && !node.has_fused_primitives()) {
+                auto& pnode = node.get_users().front()->as<permute>();
+                if (!pnode.has_fused_primitives()) {
+                    auto input_lay = pnode.get_dependency(0).get_output_layout();
+                    auto output_lay = pnode.get_output_layout();
+                    if (input_lay.compatible(output_lay)) {
+                        for (int32_t fmt_idx = format::bfyx ; fmt_idx < format::oiyx ; fmt_idx++) {
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Applied

**Review by isanghao — COMMENTED** [review]
looks good to me. let me check performance.

**Review by isanghao — APPROVED** [review]

**Review by isanghao — APPROVED** [review]

**Pull request merged**